### PR TITLE
Support limiting the number of unacknowledged splits per task

### PR DESCRIPTION
**Pull request by pettyjamesm**
Extracted changes from https://github.com/prestodb/presto/pull/15761

Adds support for tracking (and optionally limiting) the number of splits that are not yet acknowledged for a given task. Previously, the only two scheduler supported limits were the number of total splits across all tasks on the worker node and the number of splits queued for a given task which included:
- splits received by the worker but not yet running as of the last task status update
- splits queued in the coordinator local `RemoteTask` but not yet sent to the task
- splits assigned in the current scheduling batch but not yet added to the coordinator-local `RemoteTask`

The new `max_unacknowledged_splits_per_task` session property enables setting a limit on the number of splits that are either:
- queued on the coordinator-local `RemoteTask` but not yet sent or at least not confirmed to have been received by the worker
- assigned to the task in the current scheduling batch but not yet added to the coordinator-local `RemoteTask`

This limit enforcement takes precedence over both of the other existing split limit configurations and is designed to prevent large task update requests that might cause a query to fail.

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/execution/TestNodeScheduler.java [hunk 1]
```diff
@@ -333,9 +340,27 @@ public void testMaxSplitsPerNode()
...
+    {
+        setUpNodes();
+        // One split for each node, and one extra split that can't be placed
+        int splitCount = nodeManager.getActiveConnectorNodes(CONNECTOR_ID).size() + 1;
+        Set<Split> splits = new HashSet<>();
+        for (int i = 0; i < splitCount; i++) {
+            splits.add(new Split(CONNECTOR_ID, new TestSplitRemote(), Lifespan.taskWide()));
+        }
+        Multimap<InternalNode, Split> assignments = nodeSelector.computeAssignments(splits, ImmutableList.copyOf(taskMap.values()), 1).getAssignments();
+        assertEquals(assignments.entries().size(), 3);
```
Comment (dain) [inline core/trino-main/src/test/java/io/trino/execution/TestNodeScheduler.java:354]: Why use the constant 3 here and `nodeManager.getActiveConnectorNodes(CONNECTOR_ID).size()` above?

**Review by dain — COMMENTED** [review]
Looks good, just some minor comments

File: core/trino-main/src/main/java/io/trino/execution/scheduler/FixedSourcePartitionedScheduler.java [hunk 2]
```diff
@@ -262,23 +263,27 @@ public void close()
...
-                Supplier<? extends List<RemoteTask>> remoteTasks)
+                Supplier<? extends List<RemoteTask>> remoteTasks,
+                int maxUnacknowledgedSplitsPerTask)
         {
             this.nodeSelector = requireNonNull(nodeSelector, "nodeSelector is null");
             this.allNodes = ImmutableList.copyOf(requireNonNull(allNodes, "allNodes is null"));
             this.bucketNodeMap = requireNonNull(bucketNodeMap, "bucketNodeMap is null");
             this.remoteTasks = requireNonNull(remoteTasks, "remoteTasks is null");
+            this.maxUnacknowledgedSplitsPerTask = maxUnacknowledgedSplitsPerTask;
+            checkArgument(maxUnacknowledgedSplitsPerTask > 0, "maxUnacknowledgedSplitsPerTask must be > 0, found: %s", maxUnacknowledgedSplitsPerTask);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/FixedSourcePartitionedScheduler.java:280]: The logic here is that a value of `0` is also invalid. I've found formatting the observed value into the error message can be useful for debugging (ie: it's useful to know whether the value was 0, Integer.MIN_VALUE, -1, etc).

File: core/trino-main/src/test/java/io/trino/execution/TestNodeScheduler.java (hunk 1 above)
Comment (pettyjamesm) [reply core/trino-main/src/test/java/io/trino/execution/TestNodeScheduler.java]: Extracted common variable for node count to use here.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java [hunk 3]
```diff
@@ -133,7 +133,7 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
                     List<InternalNode> candidateNodes = selectExactNodes(nodeMap, split.getAddresses(), includeCoordinator);
 
                     Optional<InternalNode> chosenNode = candidateNodes.stream()
-                            .filter(ownerNode -> assignmentStats.getTotalSplitCount(ownerNode) < maxSplitsPerNode)
+                            .filter(ownerNode -> assignmentStats.getUnacknowledgedSplitCountForStage(ownerNode) < maxUnacknowledgedSplitsPerTask && assignmentStats.getTotalSplitCount(ownerNode) < maxSplitsPerNode)
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/UniformNodeSelector.java]: Reordered

File: core/trino-main/src/main/java/io/trino/execution/scheduler/NodeAssignmentStats.java [hunk 4]
```diff
@@ -20,48 +20,100 @@
...
+        String nodeId = node.getNodeIdentifier();
+        PendingSplitInfo info = stageQueuedSplitInfo.get(nodeId);
+        if (info == null) {
+            info = new PendingSplitInfo(0);
+            stageQueuedSplitInfo.put(nodeId, info);
+        }
+        return info;
+    }
+
+
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/NodeAssignmentStats.java]: Removed

File: core/trino-main/src/main/java/io/trino/execution/scheduler/NodeAssignmentStats.java [hunk 5]
```diff
@@ -20,48 +20,100 @@
...
-        return assignmentCount.getOrDefault(node, 0) + splitCountByNode.computeIfAbsent(node, nodeTaskMap::getPartitionedSplitsOnNode);
+        int nodeTotalSplits = nodeTotalSplitCount.computeIfAbsent(node, nodeTaskMap::getPartitionedSplitsOnNode);
+        PendingSplitInfo stageInfo = stageQueuedSplitInfo.get(node.getNodeIdentifier());
+        return nodeTotalSplits + (stageInfo == null ? 0 : stageInfo.getAssignedSplitCount());
     }
 
     public int getQueuedSplitCountForStage(InternalNode node)
     {
-        return queuedSplitCountByNode.getOrDefault(node.getNodeIdentifier(), 0) + assignmentCount.getOrDefault(node, 0);
+        PendingSplitInfo info = stageQueuedSplitInfo.get(node.getNodeIdentifier());
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/NodeAssignmentStats.java]: Local variables renamed to `stageInfo`

File: core/trino-main/src/main/java/io/trino/execution/scheduler/NodeAssignmentStats.java [hunk 6]
```diff
@@ -20,48 +20,100 @@
...
     public void removeAssignedSplit(InternalNode node)
     {
-        assignmentCount.merge(node, 1, (x, y) -> x - y);
+        getOrCreateStageSplitInfo(node).removeAssignedSplit();
+    }
+
+    private PendingSplitInfo getOrCreateStageSplitInfo(InternalNode node)
+    {
+        String nodeId = node.getNodeIdentifier();
+        PendingSplitInfo info = stageQueuedSplitInfo.get(nodeId);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/NodeAssignmentStats.java]: Added comment

File: core/trino-main/src/main/java/io/trino/execution/scheduler/NodeSelector.java [hunk 7]
```diff
@@ -44,7 +44,7 @@
      * If we cannot find an assignment for a split, it is not included in the map. Also returns a future indicating when
      * to reattempt scheduling of this batch of splits, if some of them could not be scheduled.
      */
-    SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks);
+    SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, int maxUnacknowledgedSplitsPerTask);
```
[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/NodeSelector.java:47]: Still investigating this. I think there was a reason I didn't do that but I can't recall what that was off the top of my head right now. Will update once I have a reason or once I've made the change.

[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/NodeSelector.java:47]: Ok, I think there were basically two reasons I did it this way:
- Only source stages respect `maxUnacknowledgedSplitsPerTask` which made it seem slightly more logical that the session property would get bound to the `SplitPlacementPolicy` (via `StageScheduler`). Binding it via the `NodeSelector` would have required passing the session (or a specific `int` value) into some strange places like ScaledWriterScheduler creation https://github.com/trinodb/trino/blob/4e51233cc1ef193de4c5e2731aa7441fc8c8a59f/core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java#L380 . This was a judgement call though can the awkward placement is going to happen either one or the other.
- Writing tests was going to be much harder if I had to re-create `NodeSelector` instances between scheduling iterations, so I chose to the path of least resistance on the judgement call above.

With that context, I'm happy to make the change to switch things over if you still think it makes sense.

[the remark this answers is not in the document]
Comment (pettyjamesm) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/NodeSelector.java]: This ended up not being too bad, so I made the change. Ready for re-review.

**Pull request merged**